sunxspex/sunxspex_fitting/stix_spec_code.py

```diff
@@ -117,20 +121,105 @@ def _get_srm_file_info(srm_file):
+    ATT_OPTS = ('attenuated', 'unattenuated')
+    if srm_choice not in ATT_OPTS:
+        raise ValueError(f"Select a STIX attenuator state from: {', '.join(ATT_OPTS)}")
+
+    with fits.open(srm_file) as sf:
+        # get units from the fits columns rather than hard-coding
+        slice_unit = lambda hdu, c: u.Unit(sf[hdu].columns[c].unit)
+
+        # changes for attenuator vs no attenuator
+        srm_idx = (1 if srm_choice == 'unattenuated' else 4)
+        photon_bins = np.column_stack((
+            sf[srm_idx].data['energ_lo'],
+            sf[srm_idx].data['energ_hi']
+        )) << slice_unit(srm_idx, 'energ_lo')
+
+        # the 1/keV part is from count bins
+        srm = sf[srm_idx].data['matrix'] << (u.ct / u.keV / u.ph)
+        count_bins = np.column_stack(
+            (sf['ebounds'].data['e_min'], sf['ebounds'].data['e_max'])
+        ) << slice_unit('ebounds', 'e_min')
+
+        # ...
+        area = sf['specresp matrix'].header['geoarea'] << u.cm**2
+
+    return SrmFileInfo(
+        photon_bin_edges=photon_bins.to_value(u.keV),
+        count_bin_edges=count_bins.to_value(u.keV),
+        srm=(srm * np.diff(count_bins, axis=1).flatten() * area).to_value(u.cm**2 * u.ct / u.ph)
+    )
+
+
+def estimate_livetime(triggers, counts_arr, time_bins, num_detectors):
+    '''
+    Functional form pulled from stixdcpy.
+    Updated time constants from correspondence with Hannah Collier.
+    '''
+    # TODO: adjust for pixel data when we need to
+    time_bins = (time_bins << u.s).value
+
+    BETA = 0.94
+    FPGA_TAU = 10.1e-6
+    # Updated ASIC_TAU from Hannah Collier (July 2023)
+    ASIC_TAU = 1.1e-6
+    TRIG_TAU = FPGA_TAU + ASIC_TAU
+    # STIX detector parameters
+    tau_conv_const = 1e-6
+
+    photons_in = triggers / (
+        time_bins * num_detectors - TRIG_TAU * triggers
+    )
+    # photon rate approximated using triggers
+
+    count_rate = counts_arr / time_bins[:, None]
+    nin = photons_in
+
+    live_ratio = np.exp(-BETA * nin * ASIC_TAU * tau_conv_const)
+    live_ratio /= (1 + nin * TRIG_TAU)
+    corrected_rate = count_rate / live_ratio[:, None]
+    return {
+        'corrected_rate': corrected_rate,
+        'count_rate': count_rate,
+        'photons_in': photons_in,
+        'live_ratio': live_ratio
+    }
+
+
+@u.quantity_input
+def compute_livetimes(
+    triggers: u.one,
+    trig_comp_err: u.one,
+    num_dets: int,
+    counts: u.ct,
+    time_bins: u.s
+):
+    '''
+    Estimate livetime and its error from triggers and counts.
+    '''
+
+    lt_high = estimate_livetime(
+        triggers=triggers + trig_comp_err,
+        counts_arr=counts,
+        time_bins=time_bins,
+        num_detectors=num_dets,
+    )
+
+    lt_low = estimate_livetime(
+        triggers=triggers - trig_comp_err,
+        counts_arr=counts,
+        time_bins=time_bins,
+        num_detectors=num_dets
+    )
+
+    lt_err = np.abs(lt_low['live_ratio'] - lt_high['live_ratio']) / 2
```

Comment:
I need to check but I don't think the compression error are symmetric and does this change this calculation?